app/assets/javascripts/govuk_publishing_components/components/single-page-notification-button.js

```diff
@@ -0,0 +1,43 @@
+/* global XMLHttpRequest */
+window.GOVUK = window.GOVUK || {}
+window.GOVUK.Modules = window.GOVUK.Modules || {};
+
+(function (Modules) {
+  function SinglePageNotificationButton ($module) {
+    this.$module = $module
+    this.personalisationEndpoint = this.$module.getAttribute('data-personalisation-endpoint')
+  }
+
+  SinglePageNotificationButton.prototype.init = function () {
+    var xhr = new XMLHttpRequest()
+    xhr.open('GET', this.personalisationEndpoint, true)
+
+    xhr.onreadystatechange = function () {
+      if (xhr.readyState === 4 && xhr.status === 200) {
+        var responseText = xhr.responseText
+        // if response text exists and is JSON parse-able, parse the response and get the button html
+        if (responseText && this.responseIsJSON(responseText)) {
+          var newButton = JSON.parse(responseText).button_html
+          var html = document.createElement('div')
+          html.innerHTML = newButton
+          // test that the html returned contains the button component; if yes, swap the button for the updated version
+          var responseHasButton = html.querySelector('form.gem-c-single-page-notification-button')
+          if (responseHasButton) {
+            this.$module.outerHTML = newButton
+          }
+        }
+      }
+    }.bind(this)
+    xhr.send()
```

Comment:
Trying to get my head around this... does this script need to take account of user cookie preferences? I can't remember how that works with regard to a user being signed in.

I think the thing I'm concerned about is that this script does an XMLHttpRequest by default, when most users (at least to start with) won't be signed into an account. Is there any way to make a check for this first before firing a network request?

Reply:
The account session cookie is an `HTTPOnly` cookie, so it's not accessible to javascript (but is still sent with `XMLHttpRequest`).  So we can't check for its presence here.

However, we *could* serve two versions of the page, with slightly different markup: one for users with a cookie and one for users without a cookie.  I think that'll work, so long as we set the right cache-control response header.

Reply:
@andysellick Michael suggested to make the JS stuff opt-in which I think is a great idea and shouldn't be a lot more additional work. Bonus: it will also appease this visual regression test which is complaining that `/api/personalisation/check-email-subscription` does not exist in the component guide https://github.com/alphagov/govuk_publishing_components/runs/4189636063?check_suite_focus=true#step:6:55

Reply:
By 'opt-in' do you mean it'll only run if the user has accepted cookies?

Reply:
> By 'opt-in' do you mean it'll only run if the user has accepted cookies?

@andysellick  What I meant about 'opt-in' here is get the component JS to run only when a user is logged in (as Michael's detailed above).
You brought up a good point about cookies however. I was not sure if the functionality should be present only when someone has accepted cookies or not. I brought up this question with team and it was concluded that the answer is no, because the only cookie this functionality is dependent on is the account session cookie, which is considered an essential one and thus does not require explicit consent.